### PR TITLE
#46 - Fixed UploadSlice

### LINKED ISSUE #46: UploadSlice does not actually saves content to the storage
Current `UploadSlice` implementation has 2 issues:
* It does not actually await for storage to complete save operation before sending OK response
* Data is not getting saved, because `Publisher` content is read twice

### PR DESCRIPTION
**Pull request by olegmoz**
Closes #46 
Fixed bug with UploadSlice not saving content & not propagating errors

### CODE REVIEW

**Comment by g4s8** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/artipie/npm-adapter/issues/47#issuecomment-615249873))

@g4s8 This is what I know about this job in CT2E6TK9B, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#47](https://github.com/artipie/npm-adapter/issues/47) is [in scope](http://www.zerocracy.com/policy.html#14) for 15min
  * The role is `REV`
  * The job is not assigned to anyone
  * The [budget](http://www.zerocracy.com/policy.html#4) is not set yet
  * These users are banned and won't be assigned:
    * @olegmoz[/z](https://www.0crat.com/u/olegmoz): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT2E6TK9B?q=%7Bjob%3A%27gh%3Aartipie%2Fnpm-adapter%2347%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by g4s8** [comment]
@0crat elect

**Comment by 0crat** [comment]
This pull request [#47](https://github.com/artipie/npm-adapter/issues/47) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/9f345e6d-5295-4eee-bc81-1d3537c97397); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artipie/npm/http/UploadSlice.java [hunk 1]
```diff
@@ -79,23 +83,29 @@ public Response response(
...
-                            "%s/-%s-%s.tgz",
-                            path, path,
-                            new LastVersion(
-                                new DescSortedVersions(
-                                    json.getJsonObject("versions")
+            new Concatenation(body).single()
+                .map(Remaining::new)
+                .map(Remaining::bytes)
+                .to(SingleInterop.get())
+                .thenCompose(
```
Comment (Vatavuk) [inline src/main/java/com/artipie/npm/http/UploadSlice.java:90]: @olegmoz let's add the compose part to a private method to make it more readable

Comment (olegmoz) [reply src/main/java/com/artipie/npm/http/UploadSlice.java]: @Vatavuk yes, that would make code easier to read. Done, thanks

**Comment by Vatavuk** [comment]
@g4s8 please remove this issue from the scope

**Comment by 0crat** [comment]
@vatavuk[/z](https://www.0crat.com/u/vatavuk) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by g4s8** [comment]
@0crat out

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @vatavuk[/z](https://www.0crat.com/u/vatavuk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@Vatavuk please make sure you will find at least 3 issues during next CR, thank

**Comment by Vatavuk** [comment]
@sereshqua ok

**Comment by sereshqua** [comment]
@0crat quality acceptable

**Pull request merged**